// @microsoft/ts-command-line/lib/test/DynamicCommandLineParser.test.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
const DynamicCommandLineParser_1 = require("../DynamicCommandLineParser");
const DynamicCommandLineAction_1 = require("../DynamicCommandLineAction");
describe('DynamicCommandLineParser', () => {
    it('parses an action', () => {
        const commandLineParser = new DynamicCommandLineParser_1.DynamicCommandLineParser({
            toolFilename: 'example',
            toolDescription: 'An example project'
        });
        const action = new DynamicCommandLineAction_1.DynamicCommandLineAction({
            actionName: 'do-job',
            summary: 'does the job',
            documentation: 'a longer description'
        });
        commandLineParser.addAction(action);
        action.defineFlagParameter({
            parameterLongName: '--flag',
            description: 'The flag'
        });
        return commandLineParser.execute(['do-job', '--flag']).then(() => {
            expect(commandLineParser.selectedAction).toEqual(action);
            const retrievedParameter = action.getFlagParameter('--flag');
            expect(retrievedParameter.value).toBe(true);
        });
    });
});
