// @microsoft/ts-command-line/lib/index.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * An object-oriented command-line parser for TypeScript projects.
 *
 * @packagedocumentation
 */
var CommandLineAction_1 = require("./CommandLineAction");
exports.CommandLineAction = CommandLineAction_1.CommandLineAction;
var CommandLineParameter_1 = require("./CommandLineParameter");
exports.CommandLineParameterKind = CommandLineParameter_1.CommandLineParameterKind;
exports.CommandLineParameter = CommandLineParameter_1.CommandLineParameter;
exports.CommandLineParameterWithArgument = CommandLineParameter_1.CommandLineParameterWithArgument;
exports.CommandLineStringParameter = CommandLineParameter_1.CommandLineStringParameter;
exports.CommandLineStringListParameter = CommandLineParameter_1.CommandLineStringListParameter;
exports.CommandLineFlagParameter = CommandLineParameter_1.CommandLineFlagParameter;
exports.CommandLineIntegerParameter = CommandLineParameter_1.CommandLineIntegerParameter;
exports.CommandLineChoiceParameter = CommandLineParameter_1.CommandLineChoiceParameter;
var CommandLineParameterProvider_1 = require("./CommandLineParameterProvider");
exports.CommandLineParameterProvider = CommandLineParameterProvider_1.CommandLineParameterProvider;
var CommandLineParser_1 = require("./CommandLineParser");
exports.CommandLineParser = CommandLineParser_1.CommandLineParser;
var DynamicCommandLineAction_1 = require("./DynamicCommandLineAction");
exports.DynamicCommandLineAction = DynamicCommandLineAction_1.DynamicCommandLineAction;
var DynamicCommandLineParser_1 = require("./DynamicCommandLineParser");
exports.DynamicCommandLineParser = DynamicCommandLineParser_1.DynamicCommandLineParser;
